source/common/grpc/transcoding_filter.h

```diff
@@ -0,0 +1,117 @@
+#include "grpc_transcoding/transcoder.h"
+#include "grpc_transcoding/type_helper.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/** VariableBinding specifies a value for a single field in the request message.
+ * When transcoding HTTP/REST/JSON to gRPC/proto the request message is
+ * constructed using the HTTP body and the variable bindings (specified through
+ * request url).
+ * See https://github.com/googleapis/googleapis/blob/master/google/api/http.proto
+ * for details of variable binding.
+ */
+struct VariableBinding {
+  // The location of the field in the protobuf message, where the value
+  // needs to be inserted, e.g. "shelf.theme" would mean the "theme" field
+  // of the nested "shelf" message of the request protobuf message.
+  std::vector<std::string> field_path;
+  // The value to be inserted.
+  std::string value;
+};
+
+/**
+ * Global configuration for the gRPC transcoding filter. Factory for the Transcoder interface.
+ */
+class TranscodingConfig : public Logger::Loggable<Logger::Id::config> {
+public:
+  /**
+   * constructor that loads protobuf descriptors from the file specified in the JSON config.
+   * and construct a path matcher for HTTP path bindings.
+   */
+  TranscodingConfig(const Json::Object& config);
+
+  /**
+   * Create an instance of Transcoder interface based on incoming request
+   * @param headers headers received from decoder
+   * @param request_input a ZeroCopyInputStream reading from downstream request body
+   * @param response_input a TranscoderInputStream reading from upstream response body
+   * @param transcoder output parameter for the instance of Transcoder interface
+   * @param method_descriptor output parameter for the method looked up from config
+   * @return status whether the Transcoder instance are successfully created or not
+   */
+  google::protobuf::util::Status
+  createTranscoder(const Http::HeaderMap& headers,
+                   google::protobuf::io::ZeroCopyInputStream& request_input,
+                   google::grpc::transcoding::TranscoderInputStream& response_input,
+                   std::unique_ptr<google::grpc::transcoding::Transcoder>& transcoder,
+                   const google::protobuf::MethodDescriptor*& method_descriptor);
+
+  /**
+   * Convert method descriptor to RequestInfo that needed for transcoding library
+   */
+  google::protobuf::util::Status
+  methodToRequestInfo(const google::protobuf::MethodDescriptor* method,
+                      google::grpc::transcoding::RequestInfo* info);
+
+private:
+  google::protobuf::DescriptorPool descriptor_pool_;
+  google::grpc::transcoding::PathMatcherPtr<const google::protobuf::MethodDescriptor*>
+      path_matcher_;
+  std::unique_ptr<google::grpc::transcoding::TypeHelper> type_helper_;
+};
+
+typedef std::shared_ptr<TranscodingConfig> TranscodingConfigSharedPtr;
+
+class TranscodingFilter : public Http::StreamFilter, public Logger::Loggable<Logger::Id::http2> {
```

Comment:
Can we get some class level comment on what this filter does? Or just comment to refer to docs (which need more detail per other comment).